echopop/computation/biology.py

```diff
@@ -104,83 +128,200 @@ def sum_strata_weight( haul_data: pd.DataFrame ,
+    return weight_strata_aged_unaged , weight_strata
 
-def sum_strata_length_age_sex_weight( haul_data: pd.DataFrame ,
-                                      specimen_data: pd.DataFrame ,
-                                      length_intervals: np.ndarray ,
-                                      age_intervals: np.ndarray ):
-    
-    ### Sum weights from each station and strata 
-    weight_strata_station , weight_strata = sum_strata_weight( haul_data ,
-                                                               specimen_data )
-    
-    ### Calculate summed length-age-sex weight bins across strata
-    weight_length_age_sex_stratum = (
+def compute_index_aged_weight_proportions( specimen_data: pd.DataFrame ,
+                                           length_intervals: np.ndarray ,
+                                           age_intervals: np.ndarray ):
+    """
+    Calculate length-age binned weight proportions    
+
+    Parameters
+    ----------
+    specimen_data: pd.DataFrame
+        Dataframe containing specimen weights
+    length_intervals: np.ndarray
+        Array containing length bins/intervals
+    age_intervals: np.ndarray
+        Array containing age bins/intervals
+    """ 
+
+    ### Process the specimen data 
+    specimen_binned = (
         specimen_data
+        # ---- Drop unaged fish 
+        # !!! TODO: pending what FEAT says, weights associated with 
+        # missing ages should be added into the 'unaged' category. 
+        # This would further mean apportioning these into `weight_strata_aged_uanged`
         .dropna( how = 'any' )
-        .loc[ lambda x: x.sex != 'unsexed' ]
-        .pipe( lambda df: pd.concat( [ df , 
-                                       df.copy( ).assign( sex = 'all' ) ] ) )
+        # ---- Remove unsexed fish
+        .loc[ lambda df: df.sex != 'unsexed' ]
+        # ---- Bin length
+        .bin_variable( length_intervals , 'length' )
+        # ---- Age bin
         .bin_variable( age_intervals , 'age' )
-        .bin_variable( length_intervals , 'length' )      
-        .groupby( [ 'stratum_num' , 'age_bin' , 'length_bin' , 'sex' ] )
-        .apply( lambda df: pd.Series( { 'summed_weight_all': df.weight.sum( ) ,
-                                        'summed_weight_adult': df.loc[ df.age > 1 ].weight.sum( ) } ) ) 
+    )
+
+    ### Sum weights within each length and age bin for each sex within each stratum
+    specimen_binned_weight = (
+        specimen_binned
+        # ---- Group weight summations across distributions of length and age 
+        # ---- for each sex within each stratum
+        .groupby( [ 'stratum_num' , 'sex' , 'length_bin' , 'age_bin' ] )
+        # ---- Sum the weights 
+        .apply( lambda df: pd.Series( { 'weight_all': df.weight.sum( ) ,
+                                        'weight_adult': df.loc[ df.age > 1 ].weight.sum( ) } ) ) 
+        # ---- Fill empty/non-existent values with 0's
+        .fillna( 0 )
         .reset_index( )
```

Comment:
For this I would likely do something like:
```python
specimen_binned_weight["weight_all"] = specimen_binned.groupby( ["stratum_num" , "sex", "length_bin", "age_bin"] )["weight"].sum()
specimen_binned_weight["weight_adult"] = specimen_binned[specimen_binned["age"] > 1].groupby( [ 'stratum_num' , 'sex' , 'length_bin' , 'age_bin' ] )["weight"].sum()

```

Reply:
@leewujung Ah, related to what we had briefly chatted about when we met on 27 March 2024 (for reference): I see the virtue in vectorizing the columns this way, although my eyes inexplicably prefer generating both columns simultaneously (i.e. via `apply` within the dataframe chain). In this particularly case, this doesn't work given how I have the dataframes designed. The dimensions for `specimen_binned` (3253, 13) doesn't match the output of `specimen_binned.groupby( ["stratum_num" , "sex", "length_bin", "age_bin"] )["weight"].sum().reset_index()` (15840, 5), which represents the sums of weights across $N_I$ x $N_\alpha$ x $N_\ell$ x $N_s$ (which yields 15840 unique combinations of all four indices). I think `specimen_binned` may, subequently, be a misleading name and should be amended because while it contains the columns with the particular `length_bin` and `age_bin` intervals, the weights themselves are not actually quantized/summed/etc. In a related commit I have tried to rename this accordingly. Otherwise, I think the setup toward this step could also be tweaked so the code isn't so reliant on pipes (and therefore easier to vectorize). Happy to discuss this further as well!

Reply:
I am not following your description, but once you have the tests I can evaluate the test function to see what you're referring to here.